builtInFunctions/esdtDataStorage.go

```diff
@@ -0,0 +1,233 @@
+	"github.com/ElrondNetwork/elrond-go-core/data/esdt"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+	"github.com/ElrondNetwork/elrond-vm-common/atomic"
+)
+
+//TODO: resolve GetESDTToken on blockchain hook
+
+type esdtDataStorage struct {
+	accounts                vmcommon.AccountsAdapter
+	globalSettingsHandler   vmcommon.ESDTGlobalSettingsHandler
+	marshalizer             vmcommon.Marshalizer
+	keyPrefix               []byte
+	flagSaveToSystemAccount atomic.Flag
+	saveToSystemEnableEpoch uint32
+}
+
+// ArgsNewESDTDataStorage defines the argument list for new esdt data storage handler
+type ArgsNewESDTDataStorage struct {
+	Accounts                vmcommon.AccountsAdapter
+	GlobalSettingsHandler   vmcommon.ESDTGlobalSettingsHandler
+	Marshalizer             vmcommon.Marshalizer
+	SaveToSystemEnableEpoch uint32
+	EpochNotifier           vmcommon.EpochNotifier
+}
+
+// NewESDTDataStorage creates a new esdt data storage handler
+func NewESDTDataStorage(args ArgsNewESDTDataStorage) (*esdtDataStorage, error) {
+	if check.IfNil(args.Accounts) {
+		return nil, ErrNilAccountsAdapter
+	}
+	if check.IfNil(args.GlobalSettingsHandler) {
+		return nil, ErrNilGlobalSettingsHandler
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, ErrNilMarshalizer
+	}
+	if check.IfNil(args.EpochNotifier) {
+		return nil, ErrNilEpochHandler
+	}
+
+	e := &esdtDataStorage{
+		accounts:                args.Accounts,
+		globalSettingsHandler:   args.GlobalSettingsHandler,
+		marshalizer:             args.Marshalizer,
+		keyPrefix:               []byte(core.ElrondProtectedKeyPrefix + core.ESDTKeyIdentifier),
+		flagSaveToSystemAccount: atomic.Flag{},
+		saveToSystemEnableEpoch: args.SaveToSystemEnableEpoch,
+	}
+	args.EpochNotifier.RegisterNotifyHandler(e)
+
+	return e, nil
+}
+
+// GetESDTNFTTokenOnSender gets the nft token on sender account
+func (e *esdtDataStorage) GetESDTNFTTokenOnSender(
+	accnt vmcommon.UserAccountHandler,
+	esdtTokenKey []byte,
+	nonce uint64,
+) (*esdt.ESDigitalToken, error) {
+	esdtData, isNew, err := e.GetESDTNFTTokenOnDestination(accnt, esdtTokenKey, nonce)
+	if err != nil {
+		return nil, err
+	}
+	if isNew {
+		return nil, ErrNewNFTDataOnSenderAddress
+	}
+
+	return esdtData, nil
+}
+
+// GetESDTNFTTokenOnDestination gets the nft token on destination account
+func (e *esdtDataStorage) GetESDTNFTTokenOnDestination(
+	accnt vmcommon.UserAccountHandler,
+	esdtTokenKey []byte,
+	nonce uint64,
+) (*esdt.ESDigitalToken, bool, error) {
+	esdtNFTTokenKey := computeESDTNFTTokenKey(esdtTokenKey, nonce)
+	esdtData := &esdt.ESDigitalToken{
+		Value: big.NewInt(0),
+		Type:  uint32(core.Fungible),
+	}
+	marshaledData, err := accnt.AccountDataHandler().RetrieveValue(esdtNFTTokenKey)
+	if err != nil || len(marshaledData) == 0 {
+		return esdtData, true, nil
+	}
+
+	err = e.marshalizer.Unmarshal(esdtData, marshaledData)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if !e.flagSaveToSystemAccount.IsSet() || nonce == 0 {
```

Comment:
This check could be made earlier. And if the flag is not set, then skip the old `RetrieveValue()` and go directly to `getESDTMetaDataFromSystemAccount`.

Reply:
the problem is in the transition - you do not know whether the metadata is in the account or in the system account.